### PR TITLE
Generate level change info for `/route` response

### PR DESCRIPTION
**Pull request by chrstnbwnkl**
# Issue

As discussed in #3650, Valhalla does not return any information about traversed levels in the route response. Hence, there is no way to properly render route shapes in multi-level contexts. This PR addresses this by adding a `level_changes` key to the route JSON response, which is a sparse array of arrays containing two elements: 

 - the index into the shape at which a new level has been reached 
 - the level as a floating point value 

A current caveat is that since level information is only stored on the edge rather than the node, a level change is only detected once the trip reaches an edge that is situated on only one level.      

## Tasklist

 - [x] Add tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [x] Update the docs with any new request parameters or changes to behavior described
 - [x] Update the [changelog](CHANGELOG.md)

### CODE REVIEW

File: src/loki/search.cc [hunk 1]
```diff
@@ -34,6 +34,7 @@ bool search_filter(const DirectedEdge* edge,
   // Note that min_ and max_road_class are integers where, by default, max_road_class
   // is 0 and min_road_class is 7. This filter rejects roads where the functional
   // road class is outside of the min to max range.
+
```
Comment (chrstnbwnkl) [inline src/loki/search.cc:37]: remove

File: src/thor/triplegbuilder.cc [hunk 2]
```diff
@@ -1052,11 +1052,14 @@ void ProcessTunnelBridgeTaggedValue(valhalla::StreetName* trip_edge_name,
  * @param  has_junction_name  True if named junction exists, false otherwise
  * @param  start_tile         The start tile of the start node
  * @param  blind_instructions Whether instructions should be generated for blind users
- *
+ * @param  prev_level         The previous level the path is on. Gets updated if the new edge is
+ *                            on a different level.
+ * @param  trip               The trip path
+ * @param  shape_index        The index of the edges first point into the final leg shape
```
Comment (chrstnbwnkl) [inline src/thor/triplegbuilder.cc:1058]: update docstring

File: src/thor/triplegbuilder.cc [hunk 3]
```diff
@@ -1069,16 +1072,15 @@ TripLeg_Edge* AddTripEdge(const AttributesController& controller,
                           const uint32_t start_node_idx,
                           const bool has_junction_name,
                           const graph_tile_ptr& start_tile,
-                          const uint8_t restrictions_idx,
-                          float elapsed_secs,
-                          bool blind_instructions) {
+                          bool blind_instructions,
+                          EdgeInfo& edgeinfo,
+                          const std::pair<std::vector<std::pair<float, float>>, uint32_t>& levels) {
```
Comment (chrstnbwnkl) [inline src/thor/triplegbuilder.cc:1077]: pass edgeinfo and levels, and while at it simplify the  signature a bit by simply passing `edge_itr`

File: src/thor/triplegbuilder.cc [hunk 4]
```diff
@@ -1190,7 +1192,7 @@ TripLeg_Edge* AddTripEdge(const AttributesController& controller,
     if (mode == sif::TravelMode::kPublicTransit) {
       // TODO(nils): get the actual speed here by passing in the elapsed seconds (or the whole
       // pathinfo)
-      speed = directededge->length() / elapsed_secs * kMetersPerSectoKPH;
+      speed = directededge->length() / edge_itr->elapsed_cost.secs * kMetersPerSectoKPH;
```
Comment (chrstnbwnkl) [inline src/thor/triplegbuilder.cc:1195]: some small changes as a result of that refactor

**Review by kevinkreiser — APPROVED** [review]

**Pull request merged**